### DISCUSSION TITLE
Persisting variables not in memory

### DISCUSSION THREAD

**Discussion by ricksearcy** [elsa-workflows/elsa-core#5220] [Q&A]
We have a need to implement our own context across the lifetime of a workflow regardless of server restarts etc. In the first activity I can set the var using context.SetVariable("myVar",myObj); but that is using the Elsa.WorkFlow.Memory.Variable extension. 

How can we set a variable in an initial activity that will persist for the lifetime of a workflow out of memory?

**Comment by glime-ai** [comment]
To set a variable that will persist for the lifetime of a workflow and survive server restarts, you should use the Workflow Instance storage provider instead of the Memory storage provider. The Workflow Instance storage provider ensures that variables are preserved alongside the workflow instance, which means they are serialized and stored in a persistent store, such as a database.

Here's how you can define a variable with the Workflow Instance storage provider in a programmatic workflow:

```csharp
// Define a workflow variable to capture some data.
var myVariable = new Variable<MyType>(defaultValue: myObj);

// Define your workflow.
var workflow = new Workflow
{
    // Register the variable with Workflow Instance storage.
    Variables = { myVariable.WithWorkflowInstanceStorage() },

    // Define the rest of your workflow...
};
```

In the above example, `myObj` is the default value for the variable, and `MyType` is the type of the object you want to store. By calling `.WithWorkflowInstanceStorage()` on the variable, you are specifying that the variable should be stored with the workflow instance.

If you are using designed workflows, you can set the storage provider to "Workflow" when defining the variable in the Variables panel of the Elsa Dashboard.

Remember that when using the Workflow Instance storage provider, you must ensure that the variable type is serializable to avoid exceptions during the persistence process.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Comment by ricksearcy** [comment]
I'm not sure that either solution works for our implementation. 

We are using workflows that are built via the designer that use a set of our custom activities. 

Solution 1 does not work because we are not defining the workflow in code. 

Solution 2 does not really work because the workflow could be built in any manner using the custom activities. 

What I need to a programmatic way set a persisted variable in the first custom activity that is not in memory. Without access to a code Workflow definition and without the dependency of setting it instance provider in the designer.

Is that possible?
